README.rst
``monoseq``
===========

``monoseq`` is a Python library for pretty-printing sequence strings using a
monospace font. It also provides a simple command line interface.

Sequences are pretty-printed in the for DNA and proteins traditional way
using blocks of letters where each line is prefixed with the sequence
position. User-specified regions are highlighted using different formatting
styles. The output format can be HTML or plaintext with optional styling
using ANSI escape codes for use in a terminal.

.. code-block:: pycon

    >>> import monoseq
    >>> sequence = 'MIMANQPLWLDSEVEMNHYQQSHIKSKSPYFPEDKHICWIKIFKAFGT' * 4
    >>> print monoseq.pprint_sequence(sequence, mode='plaintext')
      1  MIMANQPLWL DSEVEMNHYQ QSHIKSKSPY FPEDKHICWI KIFKAFGTMI MANQPLWLDS
     61  EVEMNHYQQS HIKSKSPYFP EDKHICWIKI FKAFGTMIMA NQPLWLDSEV EMNHYQQSHI
    121  KSKSPYFPED KHICWIKIFK AFGTMIMANQ PLWLDSEVEM NHYQQSHIKS KSPYFPEDKH
    181  ICWIKIFKAF GT

This screenshot shows the use of ANSI escape codes:

.. image:: doc/example-ansi.png
   :alt: Example of formatting with ANSI escape codes

Here is `an example <doc/example-html.html>`_ of ``monoseq`` used to output
HTML, and this `IPython Notebook
<http://nbviewer.ipython.org/github/martijnvermaat/monoseq/blob/master/doc/monoseq.ipynb>`_
shows how to pretty-print sequences in an IPython Notebook.

To install the latest release via PyPI using pip::

    pip install monoseq


Documentation
-------------

The `latest documentation <https://monoseq.readthedocs.org/>`_ with user
guide and API reference is hosted at Read The Docs.

You can also compile the documentation directly from the source code by
running ``make html`` from the ``doc/`` subdirectory. This requires
`Sphinx`_ to be installed.


Copyright
---------

``monoseq`` is licensed under the MIT License, see the LICENSE.rst file
for details. See the AUTHORS.rst file for a list of authors.


.. _Sphinx: http://sphinx-doc.org/


CHANGES.rst
Changelog
=========

Here you can see the full list of changes between each ``monoseq`` release.


Version 1.2.2
-------------

Release date to be decided.


Version 1.2.1
-------------

Released on May 10th 2015.

- Custom styling for `monoseq.ipynb.Seq`.


Version 1.2.0
-------------

Released on May 10th 2015.

- `monoseq.ipynb.Seq` for use in the IPython Notebook.


Version 1.1.1
-------------

Released on July 18th 2013.

- Actually support 10 annotation levels in `HtmlFormat` as documented.
- User guide and documentation for command line interface.


Version 1.1.0
-------------

Released on July 14th 2013.

- Read annotation from BED track.
- Better class names in `HtmlFormat` (``monoseq-annotation-*`` and
  ``monoseq-margin``).


Version 1.0.0
-------------

Released on July 13th 2013.

- User-definable formatting stylings. This breaks the 0.1.0 API for
  `monoseq.pprint_sequence` (the `mode` argument is superseded by `format`).
- Read sequences from a FASTA file.
- Command line interface.


Version 0.1.0
-------------

Released on July 12th 2013.

First public release.


tests/test_monoseq.py
"""
Tests for the monoseq module.
"""


from nose.tools import *


from monoseq.monoseq import (HtmlFormat, AnsiFormat, PlaintextFormat,
                             partition_range, pprint_sequence)


class TestMonoseq(object):
    """
    Tests for the monoseq module.
    """
    def test_partition_range(self):
        """
        Simple partition of range with two annotation levels.
        """
        assert_equal(partition_range(50, annotations=[[(0, 21), (30, 35)],
                                                      [(15, 32), (40, 46)]]),
                     [(0, 15, {0}),
                      (15, 21, {0, 1}),
                      (21, 30, {1}),
                      (30, 32, {0, 1}),
                      (32, 35, {0}),
                      (35, 40, set()),
                      (40, 46, {1}),
                      (46, 50, set())])

    def test_pprint_sequence(self):
        """
        Pretty-print a simple sequence in plaintext.
        """
        sequence = 'MIMANQPLWLDSEVEMNHYQQSHIKSKSPYFPEDKHICWIKIFKAFGT' * 4
        assert_equal(pprint_sequence(sequence, format=PlaintextFormat),
                     '  1  MIMANQPLWL DSEVEMNHYQ QSHIKSKSPY FPEDKHICWI KIFKAFGTMI MANQPLWLDS\n'
                     ' 61  EVEMNHYQQS HIKSKSPYFP EDKHICWIKI FKAFGTMIMA NQPLWLDSEV EMNHYQQSHI\n'
                     '121  KSKSPYFPED KHICWIKIFK AFGTMIMANQ PLWLDSEVEM NHYQQSHIKS KSPYFPEDKH\n'
                     '181  ICWIKIFKAF GT')

    def test_pprint_sequence_html(self):
        """
        Pretty-print sequence to HTML with two annotation levels.
        """
        sequence = 'MIMANQPLWLDSEVEMNHYQQSHIKSKSPYFPEDKHICWIKIFKAFGT' * 4
        annotations = [[(12, 23)], [(14, 15), (38, 39), (82, 83)]]
        assert_equal(pprint_sequence(sequence, annotations=annotations, format=HtmlFormat),
                     '  <span class="monoseq-margin">1</span>  MIMANQPLWL DS'
                     '<span class="monoseq-annotation-0">EV</span>'
                     '<span class="monoseq-annotation-1"><span class="monoseq-annotation-0">E</span></span>'
                     '<span class="monoseq-annotation-0">MNHYQ</span> '
                     '<span class="monoseq-annotation-0">QSH</span>'
                     'IKSKSPY FPEDKHIC'
                     '<span class="monoseq-annotation-1">W</span>'
                     'I KIFKAFGTMI MANQPLWLDS\n'
                     ' <span class="monoseq-margin">61</span>  EVEMNHYQQS HIKSKSPYFP ED'
                     '<span class="monoseq-annotation-1">K</span>'
                     'HICWIKI FKAFGTMIMA NQPLWLDSEV EMNHYQQSHI\n'
                     '<span class="monoseq-margin">121</span>  KSKSPYFPED KHICWIKIFK AFGTMIMANQ PLWLDSEVEM NHYQQSHIKS KSPYFPEDKH\n'
                     '<span class="monoseq-margin">181</span>  ICWIKIFKAF GT')

    def test_pprint_sequence_ansi(self):
        """
        Pretty-print sequence to plaintext with ANSI escapes with two
        annotation levels.
        """
        sequence = 'MIMANQPLWLDSEVEMNHYQQSHIKSKSPYFPEDKHICWIKIFKAFGT' * 4
        annotations = [[(12, 23)], [(14, 15), (38, 39), (82, 83)]]
        assert_equal(pprint_sequence(sequence, annotations=annotations, format=AnsiFormat),
                     '  1  MIMANQPLWL DS'
                     '\033[91mEV\033[0m'
                     '\033[1m\033[91mE\033[0m\033[0m'
                     '\033[91mMNHYQ\033[0m '
                     '\033[91mQSH\033[0m'
                     'IKSKSPY FPEDKHIC'
                     '\033[1mW\033[0m'
                     'I KIFKAFGTMI MANQPLWLDS\n'
                     ' 61  EVEMNHYQQS HIKSKSPYFP ED'
                     '\033[1mK\033[0m'
                     'HICWIKI FKAFGTMIMA NQPLWLDSEV EMNHYQQSHI\n'
                     '121  KSKSPYFPED KHICWIKIFK AFGTMIMANQ PLWLDSEVEM NHYQQSHIKS KSPYFPEDKH\n'
                     '181  ICWIKIFKAF GT')


tests/test_commands.py
"""
Tests for the commands module.
"""


from nose.tools import *


from monoseq.commands import _bed_iter, _fasta_iter


class TestCommands(object):
    """
    Tests for the commands module.
    """
    def test_fasta_iter(self):
        """
        Iterate over a multi-record FASTA file.
        """
        fasta = ('>sequence 1\n',
                 'TTACAGGCTACATTGCATGATCATTGCATGATTTACAGGCTACATTGCATGAT\n',
                 'TGCATGATTTACAGGCTACATTGCATGATCATTGCATGATTTACAGGCTACAT\n',
                 'GGCTACATTGCATGATCATTG\n',
                 '>sequence 2\n',
                 'GGCTACATTTACAGGCTACATTGCATGATCATTGCATGAGGCTACATTTACAG\n',
                 'AGGCTACATTGCATGATCATTGCA\n')
        result = list(_fasta_iter(fasta))
        expected = [('sequence 1',
                     'TTACAGGCTACATTGCATGATCATTGCATGATTTACAGGCTACATTGCATGAT'
                     'TGCATGATTTACAGGCTACATTGCATGATCATTGCATGATTTACAGGCTACAT'
                     'GGCTACATTGCATGATCATTG'),
                    ('sequence 2',
                     'GGCTACATTTACAGGCTACATTGCATGATCATTGCATGAGGCTACATTTACAG'
                     'AGGCTACATTGCATGATCATTGCA')]
        assert_equal(result, expected)

    def test_chrom_iter(self):
        """
        Iterate over a multi-chromosome BED file.
        """
        bed = ('browser position chr7:127471196-127495720\n',
               'browser hide all\n',
               'track name="ItemRGBDemo" description="Item RGB demonstration" visibility=2 itemRgb="On"\n',
               'chr7	127471196	127472363	Pos1	0	+	127471196	127472363	255,0,0\n',
               'chr7	127472363	127473530	Pos2	0	+	127472363	127473530	255,0,0\n',
               'chr7	127473530	127474697	Pos3	0	+	127473530	127474697	255,0,0\n',
               'chr7	127474697	127475864	Pos4	0	+	127474697	127475864	255,0,0\n',
               'chr8	127475864	127477031	Neg1	0	-	127475864	127477031	0,0,255\n',
               'chr8	127477031	127478198	Neg2	0	-	127477031	127478198	0,0,255\n',
               'chr2	127478198	127479365	Neg3	0	-	127478198	127479365	0,0,255\n',
               'chr2	127479365	127480532	Pos5	0	+	127479365	127480532	255,0,0\n',
               'chr2	127480532	127481699	Neg4	0	-	127480532	127481699	0,0,255\n')
        result = [(chrom, list(chrom_iter))
                  for chrom, chrom_iter in _bed_iter(bed)]
        expected = [('chr7', [(127471196, 127472363),
                              (127472363, 127473530),
                              (127473530, 127474697),
                              (127474697, 127475864)]),
                    ('chr8', [(127475864, 127477031),
                              (127477031, 127478198)]),
                    ('chr2', [(127478198, 127479365),
                              (127479365, 127480532),
                              (127480532, 127481699)])]
        assert_equal(result, expected)


monoseq/ipynb.py
"""
Convenience wrapper around ``monoseq`` for use in the IPython Notebook.

.. moduleauthor:: Martijn Vermaat <martijn@vermaat.name>

.. Licensed under the MIT license, see the LICENSE.rst file.
"""


import binascii
import os

from IPython.display import HTML

from .monoseq import HtmlFormat, pprint_sequence


#: Default CSS for styling in the IPython Notebook, suporting up to four
#: levels of annotation. They are displayed as red, inverted, underlined, and
#: bold.
DEFAULT_STYLE = """
{selector} {{
    background: white;
    color: black;
    font-weight: normal;
    text-decoration: none;
}}
{selector} .monoseq-margin {{
    color: grey;
}}
{selector} .monoseq-annotation-0 {{
    color: red;
}}
{selector} .monoseq-annotation-1 {{
    background: black;
    color: white;
}}
{selector} .monoseq-annotation-1 .monoseq-annotation-0 {{
    background: red;
    color: white;
}}
{selector} .monoseq-annotation-2 {{
    text-decoration: underline;
}}
{selector} .monoseq-annotation-3 {{
    font-weight: bold;
}}
"""


def Seq(sequence, annotations=None, block_length=10, blocks_per_line=6,
        style=DEFAULT_STYLE):
    """
    Pretty-printed sequence object that's displayed nicely in the IPython
    Notebook.

    :arg style: Custom CSS as a `format string`, where a selector for the
        top-level ``<pre>`` element is substituted for `{selector}`. See
        :data:`DEFAULT_STYLE` for an example.
    :type style: str

    For a description of the other arguments, see
    :func:`monoseq.pprint_sequence`.
    """
    seq_id = 'monoseq-' + binascii.hexlify(os.urandom(4))
    pprinted = pprint_sequence(sequence,
                               annotations=annotations,
                               block_length=block_length,
                               blocks_per_line=blocks_per_line,
                               format=HtmlFormat)

    return HTML('<style>{style}</style><pre id="{seq_id}">{pprinted}</pre>'
                .format(style=style.format(selector='#' + seq_id),
                        seq_id=seq_id,
                        pprinted=pprinted))


monoseq/__init__.py
"""
``monoseq``, a Python library for pretty-printing DNA and protein sequences
using a monospace font.

.. moduleauthor:: Martijn Vermaat <martijn@vermaat.name>

.. Licensed under the MIT license, see the LICENSE.rst file.
"""


from .monoseq import (AnsiFormat, Format, HtmlFormat, partition_range,
                      PlaintextFormat, pprint_sequence)


# We follow a versioning scheme compatible with setuptools [1] where the
# package version is always that of the upcoming release (and not that of the
# previous release), post-fixed with ``.dev``. Only in a release commit, the
# ``.dev`` is removed (and added again in the next commit).
#
# Note that this scheme is not 100% compatible with SemVer [2] which would
# require ``-dev`` instead of ``.dev``.
#
# [1] http://peak.telecommunity.com/DevCenter/setuptools#specifying-your-project-s-version
# [2] http://semver.org/


__version_info__ = ('1', '2', '1')
__date__ = '10 May 2015'


__version__ = '.'.join(__version_info__)
__author__ = 'Martijn Vermaat'
__contact__ = 'martijn@vermaat.name'
__homepage__ = 'https://github.com/martijnvermaat/monoseq'


monoseq/commands.py
"""
Command line interface to ``monoseq``.

.. moduleauthor:: Martijn Vermaat <martijn@vermaat.name>

.. Licensed under the MIT license, see the LICENSE.rst file.
"""


from __future__ import print_function

import argparse
import collections
import itertools
import sys

from .monoseq import AnsiFormat, pprint_sequence


def _until_eof(stream):
    """
    Yield lines from an open file.

    Iterating over `sys.stdin` continues after EOF marker. This is annoying,
    since it means having to type ``^D`` twice to stop. Wrap this function
    around the stream to stop at the first EOF marker.
    """
    while True:
        line = stream.readline()
        if line == '':
            break
        yield line


def _fasta_iter(fasta):
    """
    Given an open FASTA file, yield tuples of (`header`, `sequence`).
    """
    # Adapted from Brent Pedersen.
    # http://www.biostars.org/p/710/#1412
    groups = (group for _, group in
              itertools.groupby(fasta, lambda line: line.startswith('>')))
    for group in groups:
        header = next(group)[1:].strip()
        sequence = ''.join(line.strip() for line in next(groups))
        yield header, sequence


def _bed_iter(bed):
    """
    Given an open BED file, yield tuples of (`chrom`, `chrom_iter`) where
    `chrom_iter` yields tuples of (`start`, `stop`).
    """
    records = (line.split()[:3] for line in bed if
               not (line.startswith('browser') or line.startswith('track')))

    for chrom, chrom_iter in itertools.groupby(records, lambda x: x[0]):
        yield chrom, ((int(start), int(stop))
                      for _, start, stop in chrom_iter)


def _pprint_fasta(fasta, annotations=None, annotation_file=None,
                  block_length=10, blocks_per_line=6):
    """
    Pretty-print each record in the FASTA file.
    """
    annotations = annotations or []

    # Annotations by chromosome.
    as_by_chrom = collections.defaultdict(lambda: [a for a in annotations] or [])
    if annotation_file:
        for chrom, chrom_iter in _bed_iter(annotation_file):
            as_by_chrom[chrom].append(list(chrom_iter))

    for header, sequence in _fasta_iter(fasta):
        print(header)
        print(pprint_sequence(sequence,
                              annotations=as_by_chrom[header.split()[0]],
                              block_length=block_length,
                              blocks_per_line=blocks_per_line,
                              format=AnsiFormat))


def _pprint_line(line, annotations=None, annotation_file=None,
                 block_length=10, blocks_per_line=6):
    """
    Pretty-print one line.
    """
    annotations = annotations or []

    if annotation_file:
        # We just use the first chromosome defined in the BED file.
        _, chrom_iter = next(_bed_iter(annotation_file))
        annotations.append(list(chrom_iter))

    print(pprint_sequence(line, annotations=annotations,
                          block_length=block_length,
                          blocks_per_line=blocks_per_line, format=AnsiFormat))


def pprint(sequence_file, annotation=None, annotation_file=None,
           block_length=10, blocks_per_line=6):
    """
    Pretty-print sequence(s) from a file.
    """
    annotations = []

    if annotation:
        annotations.append([(first - 1, last) for first, last in annotation])

    try:
        # Peek to see if this looks like a FASTA file.
        line = next(sequence_file)
        if line.startswith('>'):
            _pprint_fasta(itertools.chain([line], sequence_file),
                          annotations=annotations,
                          annotation_file=annotation_file,
                          block_length=block_length,
                          blocks_per_line=blocks_per_line)
        else:
            _pprint_line(line.strip(), annotations=annotations,
                         annotation_file=annotation_file,
                         block_length=block_length,
                         blocks_per_line=blocks_per_line)
    except StopIteration:
        pass


def main():
    """
    Command line interface.
    """
    parser = argparse.ArgumentParser(
        description='monoseq: pretty-printing DNA and protein sequences',
        epilog='If INPUT is in FASTA format, each record is pretty-printed '
        'after printing its name and ANNOTATION (if supplied) is used by '
        'matching chromosome/record name. If INPUT contains a raw sequence, '
        'only the first chromosome in ANNOTATION is used.')
    parser.add_argument(
        'sequence_file', metavar='INPUT', nargs='?', default=sys.stdin,
        type=argparse.FileType('r'), help='file to read sequence(s) from, '
        'can be in FASTA format (default: standard input)')
    parser.add_argument(
        '-b', '--block-length', metavar='LENGTH', dest='block_length',
        type=int, default=10, help='block length in letters (default: 10)')
    parser.add_argument(
        '-l', '--blocks-per-line', metavar='BLOCKS', dest='blocks_per_line',
        type=int, default=6, help='blocks per line (default: 6)')
    parser.add_argument(
        '-a', '--annotation', metavar='POS', dest='annotation', nargs=2,
        action='append', type=int, help='first and last positions of '
        'subsequence to annotate (allowed more than once)')
    parser.add_argument(
        '-e', '--bed', metavar='ANNOTATION', dest='annotation_file',
        type=argparse.FileType('r'), help='file to read annotation from in '
        'BED format')

    args = parser.parse_args()
    pprint(_until_eof(args.sequence_file), annotation=args.annotation,
           annotation_file=args.annotation_file,
           block_length=args.block_length,
           blocks_per_line=args.blocks_per_line)


if __name__ == '__main__':
    main()


monoseq/monoseq.py
"""
``monoseq``, a Python library for pretty-printing DNA and protein sequences
using a monospace font.

.. moduleauthor:: Martijn Vermaat <martijn@vermaat.name>

.. Licensed under the MIT license, see the LICENSE.rst file.
"""


import collections
import itertools
import math


class Format(collections.namedtuple('Format', ['annotations', 'margin'])):
    """
    Type of output formats for pretty-printed sequences.

    :arg annotations: For each annotation level, a pair (`left`, `right`) of
        delimiters to use for enclosing a subsequence at that level.
    :type annotations: list
    :arg margin: A pair (`left`, `right`) of delimiters to use for enclosing
        the margin (containing sequence positions).
    :type margin: tuple

    The `annotations` field can have any number of items, any subsequent
    annotation levels will be ignored in pretty-printing sequences.
    """
    pass


#: Plaintext output format.
PlaintextFormat = Format([], ('', ''))


#: Plaintext output format with ANSI escape codes.
AnsiFormat = Format([('\033[91m', '\033[0m'),  # Red.
                     ('\033[1m', '\033[0m'),   # Bold.
                     ('\033[4m', '\033[0m')],  # Underline.
                    ('', ''))


#: HTML output format.
HtmlFormat = Format([('<span class="monoseq-annotation-%i">' % i, '</span>')
                     for i in range(10)],
                    ('<span class="monoseq-margin">', '</span>'))


def partition_range(stop, annotations=None):
    """
    Partition the range from 0 to `stop` based on annotations.

        >>> partition_range(50, annotations=[[(0, 21), (30, 35)],
        ...                                  [(15, 32), (40, 46)]])
        [(0, 15, {0}),
         (15, 21, {0, 1}),
         (21, 30, {1}),
         (30, 32, {0, 1}),
         (32, 35, {0}),
         (35, 40, set()),
         (40, 46, {1}),
         (46, 50, set())]

    :arg stop: End point (not included) of the range (similar to the `stop`
        argument of the built-in :func:`range` function).
    :type stop: int
    :arg annotations: For each annotation level, a list of (`start`, `stop`)
        pairs defining an annotated region.
    :type annotations: list

    :return: Partitioning of the range as (`start`, `stop`, `levels`) tuples
        defining a region with a set of annotation levels.
    :rtype: list

    All regions (`start`, `stop`) are defined as in slicing notation, so
    zero-based and `stop` is not included.

    The `annotations` argument is a list of annotations. An annotation is a
    list of regions as (`start`, `stop`) tuples. The level of each annotation
    is its index in `annotations`.

    Annotation regions can overlap (overlap within one level is ignored) and
    do not need to be sorted.
    """
    annotations = annotations or []

    partitioning = []
    part_start, part_levels = 0, None

    # We loop over the range, only touching positions where levels potentially
    # change.
    for p in sorted(set(itertools.chain([0, stop],
                                        *itertools.chain(*annotations)))):
        if p == stop:
            partitioning.append( (part_start, p, part_levels) )
            break

        # Annotation levels for position p.
        levels = {level for level, regions in enumerate(annotations)
                  if any(x <= p < y for x, y in regions)}

        if p == 0:
            part_levels = levels
            continue

        if levels != part_levels:
            partitioning.append( (part_start, p, part_levels) )
            part_start, part_levels = p, levels

    return partitioning


def pprint_sequence(sequence, annotations=None, block_length=10,
                    blocks_per_line=6, format=PlaintextFormat):
    """
    Pretty-print sequence for use with a monospace font.

        >>> sequence = 'MIMANQPLWLDSEVEMNHYQQSHIKSKSPYFPEDKHICWIKIFKAFGT' * 4
        >>> print pprint_sequence(sequence, format=PlaintextFormat)
          1  MIMANQPLWL DSEVEMNHYQ QSHIKSKSPY FPEDKHICWI KIFKAFGTMI MANQPLWLDS
         61  EVEMNHYQQS HIKSKSPYFP EDKHICWIKI FKAFGTMIMA NQPLWLDSEV EMNHYQQSHI
        121  KSKSPYFPED KHICWIKIFK AFGTMIMANQ PLWLDSEVEM NHYQQSHIKS KSPYFPEDKH
        181  ICWIKIFKAF GT

    :arg sequence: Sequence to pretty-print.
    :type sequence: str or any sliceable yielding slices representable as
        strings.
    :arg annotations: For each annotation level, a list of (`start`, `stop`)
        pairs defining an annotated region.
    :type annotations: list
    :arg block_length: Length of space-separated blocks.
    :type block_length: int
    :arg blocks_per_line: Number of blocks per line.
    :type blocks_per_line: int
    :arg format: Output format to use for pretty-printing. Some formats are
        pre-defined as :data:`HtmlFormat`, :data:`AnsiFormat`, and
        :data:`PlaintextFormat`.
    :type format: :class:`Format`

    :return: Pretty-printed version of `sequence`.
    :rtype: str

    All regions (`start`, `stop`) are defined as in slicing notation, so
    zero-based and `stop` is not included.

    The `annotations` argument is a list of annotations. An annotation is a
    list of regions as (`start`, `stop`) tuples. The level of each annotation
    is its index in `annotations`.

    Annotation regions can overlap (overlap within one level is ignored) and
    do not need to be sorted.

    The number of annotation levels supported depends on `format`.
    :data:`HtmlFormat` supports 10 levels, :data:`AnsiFormat` supports 3
    levels and annotations are ignored completely with
    :data:`PlaintextFormat`.
    """
    annotations = annotations or []

    partitioning = partition_range(len(sequence), annotations)

    # The maximum length for positions is the 10_log of the length of the
    # sequence.
    margin = int(math.floor(math.log(max(len(sequence), 1), 10))
                 + 1) + len(format.margin[0])
    result = (format.margin[0] + '1').rjust(margin) + format.margin[1] + ' '

    for p in range(0, len(sequence), block_length):
        # Partitioning of the block starting at position p.
        block = [(max(start, p), min(stop, p + block_length), levels)
                 for start, stop, levels in partitioning
                 if start < p + block_length and stop > p]

        result += ' '
        for start, stop, levels in block:
            delimiters = [(left, right) for level, (left, right)
                          in enumerate(format.annotations) if level in levels]
            result += (''.join(left for left, right in reversed(delimiters)) +
                       str(sequence[start:stop]) +
                       ''.join(right for left, right in delimiters))

        if (not (p + block_length) % (block_length * blocks_per_line) and
            p + block_length < len(sequence)):
            result += ('\n' + (format.margin[0] +
                               str(p + block_length + 1)).rjust(margin) +
                       format.margin[1] + ' ')

    return result
